### PR TITLE
Include additional API endpoint to support COM Api Edge usage in SMART apps

### PR DESCRIPTION
**Pull request by nagarajan-gopal**
Include additional API endpoint to support COM Api Edge usage in SMART apps

### CODE REVIEW

File: src/js/cerner-smart-embeddable-lib.js [hunk 1]
```diff
@@ -42,6 +42,12 @@ const CernerSmartEmbeddableLib = {
       CernerSmartEmbeddableLib.setFrameHeight(height);
     });
   },
+   /**
+   * COM API invocation with specified Api name and corresponding params
+   */
+   invokeComApi: function invokeComApi(Apiname, params) {
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:48]: Could we rename this method to `invokeAPI` so we can avoid sharing/substituting the type of APIs the method will call later on?

Comment (kolkheang) [reply src/js/cerner-smart-embeddable-lib.js:48]: cc: @gmodrogan

File: src/js/cerner-smart-embeddable-lib.js [hunk 2]
```diff
@@ -42,6 +42,12 @@ const CernerSmartEmbeddableLib = {
       CernerSmartEmbeddableLib.setFrameHeight(height);
     });
   },
+   /**
+   * COM API invocation with specified Api name and corresponding params
+   */
```
[the remark this answers is not in the document]
Comment (nagarajan-gopal) [reply src/js/cerner-smart-embeddable-lib.js:51]: Additional information added and updated the annotations for more clarity.

File: src/js/cerner-smart-embeddable-lib.js (hunk 1 above)
Comment (nagarajan-gopal) [reply src/js/cerner-smart-embeddable-lib.js]: Agreed with your choice for method name to be more generic

File: src/js/cerner-smart-embeddable-lib.js [hunk 3]
```diff
@@ -42,6 +42,17 @@ const CernerSmartEmbeddableLib = {
       CernerSmartEmbeddableLib.setFrameHeight(height);
     });
   },
+  /**
+   * API invocation with specified name and corresponding params.
+   * COM Api have several api's with different parameter implementation.
+   * In general it have a Api name and set of the params.
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:48]:
```suggestion
   * In general it has an API name and a set of the params.
```

File: src/js/cerner-smart-embeddable-lib.js [hunk 4]
```diff
@@ -42,6 +42,17 @@ const CernerSmartEmbeddableLib = {
       CernerSmartEmbeddableLib.setFrameHeight(height);
     });
   },
+  /**
+   * API invocation with specified name and corresponding params.
+   * COM Api have several api's with different parameter implementation.
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:47]:
```suggestion
   * -COM Api have several api's with different parameter implementation.-
```

Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:47]: I think we can remove this sentence mentioning COM API

File: src/js/cerner-smart-embeddable-lib.js [hunk 5]
```diff
@@ -42,6 +42,17 @@ const CernerSmartEmbeddableLib = {
...
+   * In general it have a Api name and set of the params.
+   * @param {string} apiName   - API name required to invoke
+   * @param {object} params - Any number of parameters that passed to API
+   */
+  invokeAPI: function invokeAPI(apiName, params) {
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:52]: Is it going to be a problem if `params` is null or undefined? Same with `apiName`.

File: src/js/cerner-smart-embeddable-lib.js [hunk 6]
```diff
@@ -42,6 +42,17 @@ const CernerSmartEmbeddableLib = {
       CernerSmartEmbeddableLib.setFrameHeight(height);
     });
   },
+  /**
+   * API invocation with specified name and corresponding params.
+   * COM Api have several api's with different parameter implementation.
+   * In general it have a Api name and set of the params.
+   * @param {string} apiName   - API name required to invoke
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:49]:
```suggestion
   * @param {string} apiName  - API name required to invoke
```

Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:49]: Shall we say, it cannot be null or empty?

File: src/js/cerner-smart-embeddable-lib.js (hunk 5 above)
Comment (nagarajan-gopal) [reply src/js/cerner-smart-embeddable-lib.js:52]: apiName can't be null, and will document it. 'params' are at least an empty object.

File: test/spec/smart-embedded-content_spec.js [hunk 7]
```diff
@@ -47,4 +47,14 @@ describe('CernerSmartEmbeddableLib', () => {
       expect(Provider.on).toHaveBeenCalledWith('iframeCustomResizer', jasmine.any(Function))
     });
   });
+  
+  describe('invokeCOMApi', () => {
+    it('trigger invokeCOMApi', () => {
+      spyOn(Provider, 'trigger');
+      /** Sample data based on cerner wiki - https://wiki.cerner.com/display/public/MPDEVWIKI/APPLINK */
+      const ApplinkParams = {'mode':100, 'launchMode': 'https://www.oracle.com', 'commandlineargs':''}
+      SmartEmbeddedContent.invokeComApi('APPLINK', ApplinkParams);
```
Comment (kolkheang) [inline test/spec/smart-embedded-content_spec.js:56]: Can you update the method name here? 
`InvokeAPI`, right?

File: test/spec/smart-embedded-content_spec.js [hunk 8]
```diff
@@ -47,4 +47,14 @@ describe('CernerSmartEmbeddableLib', () => {
       expect(Provider.on).toHaveBeenCalledWith('iframeCustomResizer', jasmine.any(Function))
     });
   });
+  
+  describe('invokeCOMApi', () => {
```
Comment (kolkheang) [inline test/spec/smart-embedded-content_spec.js:51]: Please update name here too

File: src/js/cerner-smart-embeddable-lib.js [hunk 9]
```diff
@@ -42,6 +42,15 @@ const CernerSmartEmbeddableLib = {
       CernerSmartEmbeddableLib.setFrameHeight(height);
     });
   },
+  /**
+   * API invocation with specified name and corresponding params
+   * @param {string} apiName   - API name required to invoke, not null  or undefined 
+   * @param {object} params - Any number of parameters that passed to API, not null or undefined 
+   */
+  invokeAPI: function invokeAPI(apiName, params) {
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:50]: Shall we add a basic check for `apiName` and `params` for type and null or undefined and log out an error to the console? I think it would help with the troubleshooting if there is any issues.

Also, without these params being valid, the invocation won't be successful, right?

Comment (kolkheang) [reply src/js/cerner-smart-embeddable-lib.js:50]: Are there already logs in the other components?

Comment (nagarajan-gopal) [reply src/js/cerner-smart-embeddable-lib.js:50]: I don't see any logs in other components of this repo. And also already have a null check in Mpage listener. But it's better to have a null check here to avoid trigger the events.

File: test/spec/smart-embedded-content_spec.js (hunk 8 above)
Comment (nagarajan-gopal) [reply test/spec/smart-embedded-content_spec.js]: Missed these test cases. Now changed!

File: test/spec/smart-embedded-content_spec.js (hunk 7 above)
Comment (nagarajan-gopal) [reply test/spec/smart-embedded-content_spec.js]: Updated!

File: test/spec/smart-embedded-content_spec.js [hunk 10]
```diff
@@ -47,4 +47,14 @@ describe('CernerSmartEmbeddableLib', () => {
       expect(Provider.on).toHaveBeenCalledWith('iframeCustomResizer', jasmine.any(Function))
     });
   });
+  
+  describe('invokeAPI', () => {
+    it('trigger invokeCOMApi', () => {
+      spyOn(Provider, 'trigger');
+      /** Sample data based on cerner wiki - https://wiki.cerner.com/display/public/MPDEVWIKI/APPLINK */
+      const ApplinkParams = {'mode':100, 'launchMode': 'https://www.oracle.com', 'commandlineargs':''}
```
Comment (kolkheang) [inline test/spec/smart-embedded-content_spec.js:55]: do we need a `;` at the end? Same for line 57 below

File: test/spec/smart-embedded-content_spec.js [hunk 11]
```diff
@@ -47,4 +47,14 @@ describe('CernerSmartEmbeddableLib', () => {
...
+      SmartEmbeddedContent.invokeAPI('APPLINK', ApplinkParams);
+      expect(Provider.trigger).toHaveBeenCalledWith('invokeCOMApi', {name: 'APPLINK', params: ApplinkParams})
+    });
+  });
```
Comment (kolkheang) [inline test/spec/smart-embedded-content_spec.js:59]: Could you add another test case for a random/generic API name string with an empty param object?

```
expect(Provider.trigger).toHaveBeenCalledWith('invokeCOMApi', {name: 'TESTAPINAME', params: {}})
```

File: src/js/cerner-smart-embeddable-lib.js [hunk 12]
```diff
@@ -42,6 +42,17 @@ const CernerSmartEmbeddableLib = {
...
+    if (apiName && params) {
+      // Trigger COM Api specific event 'invokeCOMApi'
+      Provider.trigger('invokeCOMApi', { name: apiName, params });
+    }
```
Comment (kolkheang) [inline src/js/cerner-smart-embeddable-lib.js:54]: @nagarajan-gopal 

We had a group discussion regarding the need to support all workflows for `APPLINK`. We should use the above method to trigger the API when embedded in an iframe. We should also support invoking the API directly from here without going through `Provider.trigger` when not embedded in iframe.

Comment (murtuzasyed) [reply src/js/cerner-smart-embeddable-lib.js:54]: @kolkheang, filling in for Naga. 

To support Standalone mode(not embedded in iframe), since ```APPLINK``` is available by default at the window/global scope we might need to rename the ```invokeCOMAPI```provided by ```CernerSmartEmbeddableLib```  to be just ```APPLINK```. In that way consumers or SMART apps won't have to make any change. Thoughts?

**Comment by kolkheang** [comment]
Closing this PR as #66 resolved the issue
